File: diagram-poc/backend/src/main/java/com/example/diagram/service/FeedbackLoopService.java
package com.example.diagram.service;

import com.example.diagram.web.dto.FeedbackLoopDtos.BoardDto;
import com.example.diagram.web.dto.FeedbackLoopDtos.NewEntryRequest;
import com.example.diagram.web.dto.FeedbackLoopDtos.NewThreadRequest;
import com.example.diagram.web.dto.FeedbackLoopDtos.ThreadDto;

/**
 * The feedback loop between the people working a diagram (sales, engineering,
 * customer, … — free-form roles, not a fixed set of actors): threads of
 * entries whose decisions drive the thread status.
 */
public interface FeedbackLoopService {

    BoardDto board(Long diagramId, String userEmail);

    ThreadDto createThread(Long diagramId, NewThreadRequest req, String userEmail);

    ThreadDto addEntry(Long threadId, NewEntryRequest req, String userEmail);
}

File: diagram-poc/frontend/src/app/core/services/feedback-loop.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { apiBaseUrl } from '../app-config';

/** One turn of a feedback loop. */
export interface FeedbackEntry {
  id: number;
  authorName: string;
  role: string;
  decision: 'comment' | 'request-changes' | 'approve' | 'close' | string;
  text: string;
  createdAt: string;
  mine: boolean;
}

/** A feedback thread (loop) on a diagram, optionally anchored to a block. */
export interface FeedbackThread {
  id: number;
  title: string;
  nodeId?: string | null;
  status: 'open' | 'changes-requested' | 'approved' | 'closed' | string;
  createdByName: string;
  createdAt: string;
  updatedAt: string;
  entries: FeedbackEntry[];
}

/** All threads + the role tags already used on this diagram (suggestions). */
export interface FeedbackBoard {
  threads: FeedbackThread[];
  roles: string[];
}

const API = apiBaseUrl();

/**
 * Feedback loop between the people working a diagram. Roles are free-form tags
 * chosen by each participant (sales / engineering / customer / anything) — the
 * backend never restricts who the actors are.
 */
@Injectable({ providedIn: 'root' })
export class FeedbackLoopService {
  constructor(private http: HttpClient) {}

  board(diagramId: number): Observable<FeedbackBoard> {
    return this.http.get<FeedbackBoard>(`${API}/diagrams/${diagramId}/feedback-loop`);
  }

  create(diagramId: number, req: { title: string; nodeId?: string | null; role: string; text: string }): Observable<FeedbackThread> {
    return this.http.post<FeedbackThread>(`${API}/diagrams/${diagramId}/feedback-loop`, req);
  }

  reply(threadId: number, req: { role: string; decision: string; text: string }): Observable<FeedbackThread> {
    return this.http.post<FeedbackThread>(`${API}/feedback-loop/${threadId}/entries`, req);
  }
}
